sound/drivers/pimidi.c

```diff
@@ -0,0 +1,1119 @@
+		instance->rx_buf = NULL;
+		instance->rx_len = 0;
+		instance->rx_status = -ETIMEDOUT;
+		instance->rx_completion = NULL;
+		mutex_unlock(&instance->comm_lock);
+		return -ETIMEDOUT;
+	}
+
+	if (CRC8_GOOD_VALUE(pimidi_crc8_table) != crc8(pimidi_crc8_table, result, sizeof(result),
+						       CRC8_INIT_VALUE))
+		return -EIO;
+
+	const char expected[4] = { 0xb7, 0x81, 0x01, 0x00 };
+
+	if (memcmp(result, expected, sizeof(expected)) != 0)
+		return -EPROTO;
+
+	u32 v = ntohl(*(uint32_t *)(result + 4));
+
+	version->hwrev = v >> 24;
+	version->major = (v & 0x00ff0000) >> 16;
+	version->minor = (v & 0x0000ff00) >> 8;
+	version->build = v & 0x000000ff;
+
+	return 0;
+}
+
+static int pimidi_read_serial(char serial[11], struct pimidi_instance *instance)
+{
+	memset(serial, 0, sizeof(char[11]));
+
+	const char cmd[4] = { 0xb2, 0x03, 0x04, 0x97 };
+
+	char result[PIMIDI_MAX_PACKET_SIZE];
+
+	memset(result, 0, sizeof(result));
+
+	DECLARE_COMPLETION_ONSTACK(done);
+
+	mutex_lock(&instance->comm_lock);
+	int err = i2c_master_send(instance->i2c_client, cmd, sizeof(cmd));
+
+	if (err < 0) {
+		mutex_unlock(&instance->comm_lock);
+		return err;
+	}
+	instance->rx_buf = result;
+	instance->rx_len = sizeof(result);
+	instance->rx_completion = &done;
+	mutex_unlock(&instance->comm_lock);
+
+	printd(instance, "Waiting for drdy");
+	wait_for_completion_io_timeout(&done, msecs_to_jiffies(1000u));
+	printd(instance, "Done waiting");
+
+	if (!completion_done(&done)) {
+		mutex_lock(&instance->comm_lock);
+		instance->rx_buf = NULL;
+		instance->rx_len = 0;
+		instance->rx_status = -ETIMEDOUT;
+		instance->rx_completion = NULL;
+		mutex_unlock(&instance->comm_lock);
+		printe(instance, "Timed out");
+		return -ETIMEDOUT;
+	}
+
+	if (CRC8_GOOD_VALUE(pimidi_crc8_table) != crc8(pimidi_crc8_table, result,
+						       (result[0] & 0x0f) + 2, CRC8_INIT_VALUE))
+		return -EIO;
+
+	const char expected[4] = { 0xbd, 0x83, 0x04, 0x0a };
+
+	if (memcmp(result, expected, sizeof(expected)) != 0) {
+		printe(instance, "Unexpected response: %02x %02x %02x %02x", result[0], result[1],
+		       result[2], result[3]);
+		return -EPROTO;
+	}
+
+	memcpy(serial, result + 4, 10);
+
+	if (strspn(serial, "\xff") == 10)
+		strscpy(serial, "(unset)", 8);
+
+	return 0;
+}
+
+static void pimidi_handle_midi_data(struct pimidi_instance *instance, int port, const uint8_t *data,
+				    unsigned int n)
+{
+	printd(instance, "Handling MIDI data for port %d (%u bytes)", port, n);
+	if (n == 0)
+		return;
+
+	struct pimidi_midi_port *midi_port = &instance->midi_port[port];
+
+	kfifo_in(&midi_port->in_fifo, data, n);
+
+	if (!instance->stopping)
+		queue_work(pimidi_global.work_queue, &midi_port->in_handler);
+
+	printd(instance, "Done");
+}
+
+static void pimidi_drdy_continue(struct pimidi_instance *instance)
+{
+	if (instance->stopping) {
+		printd(instance, "Refusing to queue work / enable IRQ due to stopping.");
+		return;
+	}
+
+	if (gpiod_get_value(instance->data_ready_gpio)) {
```

Comment:
Deciding whether to queue work or enable an interrupt based on the current state of the GPIO feels racy. Are you satisfied that it is safe? I guess it depends on whether the data ready line can spontaneously go high.

 I also notice that you are taking care to release the comm_lock mutex before calling this function - why?

Reply:
The DRDY line remains high as set by the Pimidi device until all of its output is exhausted, this is a safe check.

The comm mutex is released to let the kernel module communicate with other Pimidi instances.